manta-trusted-setup/src/groth16/test.rs

```diff
@@ -0,0 +1,314 @@
+        let mut hasher = BlakeHasher::new();
+        proving_key
+            .serialize(&mut hasher)
+            .expect("Hasher is not allowed to fail");
+        into_array_unchecked(hasher.0.finalize())
+    }
+}
+
+impl Types for Test {
+    type State = State<Test>;
+    type Challenge = [u8; 64];
+    type Proof = Proof<Test>;
+}
+
+/// Conducts a dummy phase one trusted setup.
+pub fn dummy_phase_one_trusted_setup() -> Accumulator<Test> {
+    let mut rng = OsRng;
+    let accumulator = Accumulator::default();
+    let challenge = [0; 64];
+    let contribution = Contribution::gen(&mut rng);
+    let proof = contribution.proof(&challenge, &mut rng).unwrap();
+    let mut next_accumulator = accumulator.clone();
+    next_accumulator.update(&contribution);
+    Accumulator::verify_transform(accumulator, next_accumulator, challenge, proof).unwrap()
+}
+
+/// Generates a dummy R1CS circuit.
+pub fn dummy_circuit(cs: &mut R1CS<Fp256<FrParameters>>) {
+    let a = Fp(field_new!(Fr, "2")).as_known::<Secret, FpVar<_>>(cs);
+    let b = Fp(field_new!(Fr, "3")).as_known::<Secret, FpVar<_>>(cs);
+    let c = &a * &b;
+    let d = Fp(field_new!(Fr, "6")).as_known::<Public, FpVar<_>>(cs);
+    c.enforce_equal(&d)
+        .expect("enforce_equal is not allowed to fail");
+}
+
+/// Proves and verifies a dummy circuit with proving key `pk` and a random number generator `rng`.
+pub fn dummy_prove_and_verify_circuit<P, R>(pk: ProvingKey<P>, mut rng: &mut R)
+where
+    P: PairingEngine<Fr = Fp256<ark_bls12_381::FrParameters>>,
+    R: CryptoRng + RngCore + ?Sized,
+{
+    let mut cs = R1CS::for_proofs();
+    dummy_circuit(&mut cs);
+    let proof = Groth16::prove(&pk, cs, &mut rng).unwrap();
+    assert!(
+        Groth16::verify(&pk.vk, &[field_new!(Fr, "6")], &proof).unwrap(),
+        "Verify proof should succeed."
+    );
+}
+
+/// Tests if proving and verifying ratio proof is correct.
+#[test]
+pub fn proving_and_verifying_ratio_proof_is_correct() {
+    let mut rng = OsRng;
+    let delta = <Test as Pairing>::Scalar::rand(&mut rng);
+    let proof = RatioProof::prove(
+        &<Test as kzg::Configuration>::hasher(Test::TAU_DOMAIN_TAG),
+        &[0; 64],
+        &delta,
+        &mut rng,
+    )
+    .expect("Proving a ratio proof should be correct.");
+    proof
+        .verify(
+            &<Test as kzg::Configuration>::hasher(Test::TAU_DOMAIN_TAG),
+            &[0; 64],
+        )
+        .expect("Verifying a ratio proof should be correct.");
+}
+
+/// Tests if trusted setup phase 2 is valid with trusted setup phase 1 and proves and verifies a
+/// dummy circuit.
+#[test]
+pub fn trusted_setup_phase_two_is_valid() {
+    let mut rng = OsRng;
+    let mut cs = R1CS::for_contexts();
+    dummy_circuit(&mut cs);
+    let accumulator = dummy_phase_one_trusted_setup();
+    let mut state =
+        initialize::<Test, R1CS<Fp256<ark_bls12_381::FrParameters>>>(accumulator, cs).unwrap();
```

Comment:
Is there a better type name than `Fp256<ark_bls12_381::FrParameters>`? See similar comment above.